### ISSUE TITLE
Non-Pypi/Pip Ruff distribution for Windows

### ISSUE THREAD

**Issue by swaldhoer** [astral-sh/ruff#8834]
# Non-Pypi/Pip Ruff distribution for Windows

## Problem to be solved

``Ruff`` is a standalone executable on all platforms and does not require Python to be installed.
For Linux and macOS ``Ruff`` is then also installable without requiring a Python/pip through the expected package managers of the OS, e.g., with Homebrew, pacman.

On Windows, Chocolatey is supported. I do not know how the build recipe for the Chocolatey package works, but anyhow I think the following suggestion would improve the developer experience on Windows for shipping and adapting ``Ruff``.

## Abstract Idea

``Ruff`` should provide some typical Windows installer, e.g., MicroSoft Installer (MSI), Nullsoft Scriptable Install System (NSIS) or just a zip file with the executable.
If the project provides this, any of the the chosen installer methods (MSI, NSIS, zip) could be used for the common Windows package managers (Chocolatey, winget, even conda - although  that use case is of course nonsense) to package ``Ruff`` without requiring Python to be installed.

This would have the great benefit that Windows users

- could have one global ``Ruff`` installation. And it just works independently of the currently active Python environment.
- could have the side-by-side-version thing - see the exact suggestion below. Everything could be very simply reproducible, as different versions can be used side-by-side.
- do not need to create an additional venv/conda environment for ``Ruff`` or at least don't need to make the make dependency list of the project longer than needed - we all know how annoying long dependency lists in the Python ecosystem are. Of course, it's just one non-Python dependency, but anyhow, the less the better.

## Actual Plan/Suggestion

- The project should provide a ruff installer for Windows.
- This installer should not require elevate privileges.
  - Running the installer as user should, e.g., install to: ``%LOCALAPPDATA%\Programs\ruff\``
  - Running the installer as administrator should, e.g., install to: ``%PROGRAMFILES%\ruff\``
- This installer should support side-by-side installations ``ruff.exe`` pointing to the latest version, e.g., for a user installation as follows

  ```pwsh
  C:\Users\user\AppData\Local\Programs\ruff
  ├───ruff.exe
  ├───ruff-0.0.1.exe
  ├───# and so on
  └───ruff-0.0.6.exe # same as 'ruff.exe'
  ```

**Comment by T-256** [comment]
> or just a zip file with the executable.

Do you mean what's provided in [GH Releases](https://github.com/astral-sh/ruff/releases)?

**Comment by swaldhoer** [comment]
Ah I didn't see that - thanks.

We could then provide an Inno Setup installer and build the installer as follows

```pwsh
cd /build/dir/where/ruff.exe/exists
copy path/to/repo/LICENSE LICENSE
ISCC.exe ruff.iss /DVersion=0.1.6
```

It installs to 
- ``%LOCALAPPDATA%\Programs\Ruff`` (user changeable)
- ``%PROGRAMFILES%\Ruff`` (user changeable)

If I polish this up, is this something you would be willing to provide to users?

<details>
  <summary>ruff.iss</summary>

```ini
; Script generated by the Inno Setup Script Wizard.
; SEE THE DOCUMENTATION FOR DETAILS ON CREATING INNO SETUP SCRIPT FILES!

#define AppName "Ruff"
#define Publisher "ASTRAL SOFTWARE INC."
#define Url "https://astral.sh/ruff"
#define ExecutableBaseName "ruff"
#define ExecutableExtension "exe"
#define Executable ExecutableBaseName + "-" + Version + "." + ExecutableExtension

[Setup]
; NOTE: The value of AppId uniquely identifies this application. Do not use the same AppId value in installers for other applications.
; (To generate a new GUID, click Tools | Generate GUID inside the IDE.)
AppId={{3C4AAAD5-ACCF-4438-A678-8C97F13B022C}
AppName={#AppName}
AppVersion={#Version}
AppVerName={#AppName} {#Version}
AppPublisher={#Publisher}
AppPublisherURL={#Url}
AppSupportURL={#Url}
AppUpdatesURL={#Url}
ArchitecturesInstallIn64BitMode=x64
DefaultDirName={autopf}\{#AppName}
DefaultGroupName={#AppName}
DisableProgramGroupPage=yes
LicenseFile=LICENSE
PrivilegesRequired=lowest
PrivilegesRequiredOverridesAllowed=dialog
OutputBaseFilename={#ExecutableBaseName}-{#Version}-installer
Compression=lzma
SolidCompression=yes
WizardStyle=modern
SetupIconFile={#ExecutableBaseName}.ico

[Languages]
Name: "english"; MessagesFile: "compiler:Default.isl"

[Files]
Source: "{#Executable}"; DestDir: "{app}"; Flags: ignoreversion

[UninstallDelete]
Type: files; Name: "{app}\{#ExecutableBaseName}.{#ExecutableExtension}"

[Icons]
Name: "{group}\{#AppName}"; Filename: "{app}\{#Executable}"
Name: "{group}\{cm:UninstallProgram,{#AppName}}"; Filename: "{uninstallexe}"

[Run]
Filename: "{cmd}"; Parameters: "/C copy ""{app}\{#Executable}"" ""{app}\{#ExecutableBaseName}.{#ExecutableExtension}"""

```
</details>

**Comment by T-256** [comment]
If you want to Ruff to be accessible globally, I'd recommend add it to PATH environment. Then you can also use it with all system users.

I think Ruff is cross-platform Python tool not OS application, so I don't think it will add support OS-Installer packaging.

**Comment by swaldhoer** [comment]
> If you want to Ruff to be accessible globally, I'd recommend add it to PATH environment. Then you can also use it with all system users.

I never said, that it is about global accessibility of ``Ruff``. It's about the way ``Ruff`` is distributed across different platforms and how it should be improved on Windows.

Anyhow, your suggestion *adding it to PATH* is currently not a real solution. Why? Because there is no standard way ``Ruff`` is distributed on Windows other than PyPI. **That is exactly the point for this issue**. What should the user *xy* add to PATH?

- ``C:\Users\xy\Ruff`` because he copied the executable there?
- ``C:\Users\xy\Miniconda3\Scripts`` because he installed it in his conda base environment (bad for several reasons, but anyhow) and so clutters his whole path with other stuff?
- ``path/to/some/repo\.venv\Scripts`` because he works on this project that requires this ``Ruff`` version? What does he do for other projects?
- ``path/to/some/other/.venv\Scripts`` because...because every Python developer has fifty environments and that's the magic one?

I could go on forever, I think my point is clear.

> I think Ruff is cross-platform Python tool, not OS application, so I don't think it will add support OS-Installer packaging.

I don't see, how your comment is related to my suggestion.
My suggestion, that there is a standard non-PyPI distribution of ``Ruff`` on Windows, is exactly what is done on other platforms in their standard version. You can check that in the documentation[^1]. I'll attach a screenshot below of the current status for the different packaging solutions[^2].

> But I agree that documentation could be improved at here. because ``Ruff`` is python independent, so different python envs could target same ``Ruff`` without installing it via pip.

Again, this thing is **not** about documentation, it is about the distribution.

I emphasize my point again:
Currently, there is **no other installation method** (we are talking about installation - not copying things around) for ``Ruff`` on Windows than PyPI. This is bad for several reasons (e.g., you should work with environments when working with Python. Why should I install ``Ruff`` in every environment? Typically I don't need 50 copies of the same executable across my machine.).

I want to have a standard way you can (silently and unattended) install ``Ruff`` on Windows in an OS-standard directory not related to Python specific directories/directory layouts (i.e., ``Ruff`` shall be placed in ``%LOCALAPPDATA%\Programs\Ruff`` or ``%PROGRAMFILES%\Ruff``). This installer should be supported by the tool maintainer. This way other tools like e.g., IDEs could rely on these paths when searching whether ``Ruff`` is available and in what version.

Why should this installer (and by that its default installation directories) maintained by the tool maintainer? So that there is an official, documented way to do it, and others can point to it when they request other tools to rely on this behavior. This is even more true on Windows as there are no derivatives (like .e.g, there are/could be on Linux). If ``Ruff`` decides, this is the way to distribute it on Windows (aside from PyPI), then we have **one**, **true** solution and everyone can work and be happy with that.

[^1]: <https://docs.astral.sh/ruff/installation>

[^2]: Packaging Status<br>![ruff-python-linter](https://github.com/astral-sh/ruff/assets/34184299/0f86152e-f363-45ea-bf6f-4084dfbf3f33)

**Comment by T-256** [comment]
Did you try Chocolatey?

**Comment by T-256** [comment]
I think you are looking for `choco install ruff` on Windows. Did you try that?

**Comment by swaldhoer** [comment]
> I think you are looking for `choco install ruff` on Windows. Did you try that?

I know (and even use) Chocolatey, but no, I am not looking for Chocolatey. You are really missing the point of this discussion: It's about the OS-native distribution and installation of ``Ruff``.

Regarding your Chocolatey comment:

- First, this is third party supported, and not by the tool maintainer/vendor. They can/could do however they think the packaging for ``Ruff`` is correct, whether ``Ruff`` likes this or not (side note: and of course they are allowed to do it).
- Second, and way more important, Chocolatey is not something that is natively supported. It's just an additional package manager. I don't want to complicate the installation for others (by requiring the installation of a package manager and then finding out how it works), I want to simplify it. Anyhow, if we provide an Inno Script, that installer could be consumed also by Chocolatey and Chocolatey would then automatically adhere to the ``Ruff`` supported way of installation ``Ruff`` on Windows.

**Comment by konstin** [comment]
Generally, we recommend to make ruff a dependency of your project so the ruff version is tracked together with the project, ensuring that every developers machine and CI are in sync. I agree that it's annoying that having ruff as a dependency on n projects installs the same file n times, but that is something that i believe should be fixed on the packaging side, since it affects all python packages, not just ruff (the problem is much bigger when e.g. installing torch multiple times).

If you still want a global installation, have you considered [pipx](https://pypa.github.io/pipx/)? It's the de-facto standard for installing python tools and applications system wide. Unlike native installers, it supports more features including  multi versions (by using prefixes such as `black@23.10.0`), updating tools and skipping installation `pipx run ruff .`.

**Comment by swaldhoer** [comment]
> Generally, we recommend to make ruff a dependency of your project so the ruff version is tracked together with the project, ensuring that every developers machine and CI are in sync. I agree that it's annoying that having ruff as a dependency on n projects installs the same file n times, but that is something that i believe should be fixed on the packaging side, since it affects all python packages, not just ruff (the problem is much bigger when e.g. installing torch multiple times).

- If I pin in my CI

  ```bash
  pip install ruff==a.b.c
  ```

  or

  ```bash
  curl https://github.com/astral-sh/ruff/releases/download/va.b.c/ruff-x86_64-pc-windows-msvc-installer.exe
  ruff-x86_64-pc-windows-msvc-installer.exe
  ```

  does not really make a difference.
- Yes, but e.g., for torch does not working without Python (because of its package requirements), so the workflow is obvious and basically the only way to use it meaningful.

> If you still want a global installation, have you considered [pipx](https://pypa.github.io/pipx/)? It's the de-facto standard for installing python tools and applications system wide. Unlike native installers, it supports more features including multi versions (by using prefixes such as `black@23.10.0`), updating tools and skipping installation `pipx run ruff .`.

But why would I want to have the overhead of a Python environment (doesn't matter pipx, conda venv, poetry, ...), when I just want to use a native executable?

Anyhow, I wonder why are you providing - let's call it - "global" non-python installers for Linux and macOS, but adding it feature parity for Windows is a debatable thing?

**Comment by swaldhoer** [comment]
Thanks for your feedback.

Anyhow, I still think one major plus of ruff is, that it is for the Python eco system, without the need of Python. Based on my experience people will **not** update their complex venv/conda/poetry/what soever environment, just to get the latest (potentially better) static code analysis tool. An installer, that adds a single binary somewhere on the system, would solve this issue and would make it easy to always use the latest version.
